core/data-layer/src/data_layer.rs

```diff
@@ -696,6 +731,121 @@ impl DataLayer {
+            + (now_ulid.get_heap_size() * 2);
+
+        let mut total_source_metrics_size = *source_metrics_size + source_metric_insert_data_size;
+
+        // get remove target data
+        let mut remove_target_data: Vec<(String, String, usize)> = Vec::new();
+        let mut subtract_total_size = total_source_metrics_size;
+        loop {
+            // check size :: buffersize - total size < 0 (None) => remove target data
+            if metric_buffer_size_kb
+                .checked_sub(subtract_total_size as u64)
+                .is_none()
+            {
+                let Some(front_source_metrics_metadata) = source_metrics_metadata.pop_front() else {
+                        break;
+                    };
+                subtract_total_size -= front_source_metrics_metadata.2; // oldest metadata size
+                remove_target_data.append(&mut vec![front_source_metrics_metadata]);
+                continue;
+            }
+            break;
+        }
+
+        // remove target data
+        remove_target_data
+            .iter()
+            .for_each(|(metric_id, ulid, size)| {
+                let Some(metric_id_data) = source_metrics.get_mut(metric_id) else {
+                    return;
+                };
+                // remove source metrics - btreemap ulid
+                metric_id_data.remove(ulid);
+                // (if btreemap is empty) remove source metrics - hashmap metric_id
+                if metric_id_data.is_empty() {
+                    source_metrics.remove(metric_id);
+                }
+                total_source_metrics_size -= size;
+            });
+
+        // add source metrics
+        match source_metrics.get_mut(metric_id) {
+            Some(source_metrics_map) => {
+                source_metrics_map.insert(now_ulid.clone(), source_metric_insert_data.clone());
+            }
+            None => {
+                let mut metric_id_map = BTreeMap::new();
+                metric_id_map.insert(now_ulid.clone(), source_metric_insert_data.clone());
+                source_metrics.insert(metric_id.to_string(), metric_id_map);
+            }
+        }
+        // add source metrics metadata
+        source_metrics_metadata.push_back((
+            metric_id.to_string(),
+            now_ulid,
+            source_metric_insert_data_size,
+        ));
+
+        debug!(
+            "[source metrics] add item size: {} / total size: {}",
+            source_metric_insert_data_size, total_source_metrics_size
+        );
+
+        // update source_metric_size
+        *source_metrics_size = total_source_metrics_size;
+
+        debug!("Save Metric\n{:?}", source_metric_insert_data);
+        Ok(())
+    }
+
+    pub async fn get_source_metrics_in_data_layer(
+        &self,
+    ) -> Result<HashMap<String, BTreeMap<String, SourceMetrics>>> {
+        let source_metrics = self.source_metrics_data.source_metrics.read().await;
+        Ok(source_metrics.clone())
+    }
```

Comment:
do not clone the original data structure that could occupy a significant amount of memory